File: HIndexII.py
class Solution:
    # Solution 1 
    # Trail H-index ==> How many papers have citations >= ith paper
    # h-index != any value in array
    # When h-index = 2 ==> Do we have atleast 2 papers >= 2 citations ?
    # Time Complexity - O(N)
    # Space Complexity - O(1)
    def hIndex1(self, citations):
        n = len(citations)
        for idx, c in enumerate(citations):
            hIndex = n - idx
            if c >= hIndex:
                return hIndex
        return 0
    
    # Using binary search by calculating trail h-index
    # Time Complexity - O(logN)
    # Space Complexity - O(1)
    def hIndex2(self, citations):
        l = len(citations)
        start = 0
        end = l - 1
        while start <= end:
            mid = start + (end-start) // 2
            trail = l - mid
            if citations[mid] == trail:
                return trail
            elif trail > citations[mid]:
                start = mid + 1
            else:
                end = mid - 1
        return l - start


        

File: h-indexII.cpp
//time complexity:O(logn)
//space complexity:O(1)
//executed on leetcode: yes
//approach:using binary search
//any issues faced? yes, the language of ques is confusing

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int l=0;
        int r=citations.size()-1;
        int n=citations.size();
        while(l<=r)
        {
            int mid=l+(r-l)/2;
            if(citations[mid]==n-mid)
                return n-mid;
            else if(citations[mid]<n-mid)
                l=mid+1;
            else
                r=mid-1;
                
        }
        return n-l;
    }
};